### ISSUE TITLE
Prefer alias instead of column for ORDER BY after GROUP BY with (prefer_column_name_to_alias=1)

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#24237]
**Describe the issue**
Prefer alias for ORDER BY after GROUP BY in case of  `set prefer_column_name_to_alias=1;`

**How to reproduce**
Clickhouse version 21.6

```
SELECT
    max(number) AS number,
    sum(number) AS sum
FROM numbers(100)
ORDER BY number ASC

Query id: 058b2cb1-7ea6-44de-b6a8-662ac4d949a8


0 rows in set. Elapsed: 0.002 sec.

Received exception from server (version 21.6.1):
Code: 184. DB::Exception: Received from localhost:9000. DB::Exception: Aggregate function max(number) is found inside another aggregate function in query: While processing max(number) AS number.

set prefer_column_name_to_alias=1;

SELECT
    max(number) AS number,
    sum(number) AS sum
FROM numbers(100)
ORDER BY number ASC

Query id: f8ca2417-9bba-42c6-9f12-42987a26e75d


0 rows in set. Elapsed: 0.004 sec.

Received exception from server (version 21.6.1):
Code: 215. DB::Exception: Received from localhost:9000. DB::Exception: Column `number` is not under aggregate function and not in GROUP BY: While processing number ASC.
```

**Comment by amosbird** [comment]
I did it at first https://github.com/ClickHouse/ClickHouse/pull/22044/files#diff-858ec3896af7064c6c5ef2c2807a7da65d5f2e793d8fb00763b0fe54982f5ed9R141

Then I reverted them https://github.com/ClickHouse/ClickHouse/pull/24022

Now I'm confused what's the expected behavior.

**Comment by UnamedRus** [comment]
> Now I'm confused what's the expected behavior.

Hmm, expected behavior for me is:
for HAVING and ORDER BY if there is aggregate function around name use column, if there is no aggregate function use alias.

**Comment by UnamedRus** [comment]
Can be included in https://github.com/ClickHouse/ClickHouse/issues/23194

**Comment by amosbird** [comment]
I think I'll add all possible options via this setting.

`set prefer_column_name_to_alias=1` --> prefer column over alias in all sub clauses;

`set prefer_column_name_to_alias=2` --> prefer column over alias in all sub clauses before and include group by;

`set prefer_column_name_to_alias=3` --> prefer column over alias in all sub clauses before group by;

**Comment by amosbird** [comment]
Could you check if https://github.com/ClickHouse/ClickHouse/pull/24391 helps resolve the issue? Please also help test other possible cases.

**Comment by alexey-milovidov** [comment]
This issue won't be addressed before #23194.
